rfcs/default_idp_configurable/default_idp_configurable.md

```diff
@@ -0,0 +1,393 @@
+    ...
+  }
+  ...
+}
+```
+
+#### KeycloakAuthenticationProviderService
+We will implement KeycloakAuthenticationProviderService for keycloak, the "validateToekn" and "revokeToken" method will be used in runtime process to validate id_token or revoke id_token. The "revokeToken" is the similar with Cognito User Pool, just need to mark this token as a revoked token, during the token validation process, if the token is already marked as revoked, the token will be validated as a an invalid token.
+Below is the code snippets keycloak token validator:
+```
+const _ = require('lodash');
+const jwt = require('jsonwebtoken');
+const jwksClient = require('jwks-rsa');
+
+async function getKeycloakTokenVerifier(keycloakUrl, keycloakRealm) {
+
+  const keyClient = jwksClient({
+    cache: true,
+    cacheMaxAge: 86400000, //value in ms
+    rateLimit: true,
+    jwksRequestsPerMinute: 10,
+    strictSsl: true,
+    jwksUri: `${keycloakUrl}/auth/realms/${keycloakRealm}/protocol/openid-connect/certs`
+  })
+  
+  const verificationOptions = {
+    "algorithms": "RS256"
+  }
+
+  function getSigningKey (header = decoded.header, callback) {
+    keyClient.getSigningKey(header.kid, function(err, key) {
+      const signingKey = key.publicKey || key.rsaPublicKey;
+      callback(null, signingKey);
+    })
+  }  
+
+  const verify = async token => {
+    // First attempt to decode token before attempting to verify the signature
+    let decoded_output;
+    var promise = new Promise(function(resolve, reject) {
+      jwt.verify(token, getSigningKey, verificationOptions, function (error, decoded) {
+        if (error) {
+          resolve(error);
+        } else {
+          resolve(decoded);
+        }
+      });
+    });
+    await promise.then(function(data) {
+      decoded_output = data;
+    });
+    return decoded_output;
+  };
+  return { verify };
+}
+
+module.exports = { getKeycloakTokenVerifier };
+```
+
+### CreateRootUserService
+Run keycloak or cognito logic depended on defaultIDPType config. Code as below in addons/addon-base-post-deployment/packages/base-post-deployment/lib/steps/create-root-user-service.js file:
+```
+  async execute() {
+    const defaultIDPType = this.settings.get(settingKeys.defaultIDPType);
+    switch (defaultIDPType) {
+      case 'Keycloak':
+        await this.createKeycloakNativeAdminUser();
+        break;
+      case 'Cognito':
+        await this.createCognitoNativeAdminUser();
+        break;
+      default:
+        console.log(`Sorry, we are out of ${defaultIDPType}.`);
+    }    
+  }
+``` 
+
+The difference with createCognitoNativeAdminUser(), createKeycloakNativeAdminUser does not need to create user in Keycloak, just need to create root user in Service Workbench DynamoDB, the user in Keycloak should be created before deploying Service Workbench. We add some configuration items in main/config/settings/${STAGE_NAME}.yml to configure root user information.
+```
+rootUserEmail: xxx
```

Comment:
Looks like those user configurations are only for KeyCloak, add prefix `keyCloak` is better

Reply:
As I explained in previous conversation, these configurations are used for both keycloak and cognito.